plugins/experimental/carp/CarpHashAlgorithm.cc

```diff
@@ -0,0 +1,396 @@
+#include <stdio.h>
+#include <memory.h>
+#include <list>
+
+#include <iostream>
+#include <sstream>
+
+#include "Common.h"
+#include "CarpHashAlgorithm.h"
+
+using namespace std;
+
+/*****************************************************************/
+void
+HashNode::dump(string& s)
+{
+  stringstream ss;
+  string sSockaddr;
+  getStringFromSockaddr(reinterpret_cast<const struct sockaddr *>( &forwardAddr), sSockaddr);
+  
+  ss << scheme << "://" << name << ":"<<listenPort<<" ("<<sSockaddr<<") weight:"<<weight<< (_status ?  string(" UP ") : string(" DOWN ")); 
+  if(_statusTime) {
+    ss << "(" << time(NULL)-_statusTime<< "s ago in "<< _statusLatencyMs << "mS)";
+  }
+  ss << " hits:" << _hits;
+  ss << " carp_noforwarded:" << _carp_noforwarded;
+  ss << " carp_forwarded:" << _carp_forwarded;
+  ss << endl;
+  s += ss.str();
+}
+
+/*****************************************************************/
+void
+HashAlgorithm::addHost(std::string name, unsigned int port, std::string scheme, double weight, bool self, struct sockaddr_storage fwdAddr)
+{
+  HashNode* node = new HashNode(name, port, scheme, weight, self, fwdAddr);
+  TSAssert(NULL != node);
+  addHost(node);
+}
+
+/*****************************************************************/
+void
+HashAlgorithm::addHost(HashNode* node)
+{
+  if (!node) return;
+  _hostList.push_back(node);
+}
+
+/*****************************************************************/
+HashNode*
+HashAlgorithm::findStatusByNameAndPort(const string& name, unsigned int port,
+    size_t* index) {
+  /*
+   * Todo: This use loop to find the corresponding HashNode
+   * 		But the HttpClient and the Hash was related, so we
+   * 		could use more easier method to write the status
+   */
+  for (size_t ptr = 0; ptr < _hostList.size(); ptr++) {
+    if (_hostList[ptr]->listenPort == port
+        && _hostList[ptr]->name.compare(name) == 0) { // found it
+      if (index) {
+        *index = ptr;
+      }
+      return _hostList[ptr];
+    }
+  }
+  return NULL;
+}
+
+size_t
+HashAlgorithm::findHashNodeIndex(HashNode *node) {
+  for (size_t ptr = 0; ptr < _hostList.size(); ptr++) {
+    if ( _hostList[ptr] == node) {
+      return ptr;
+    }
+  }
+  return -1;
+}
+
+/*****************************************************************/
+void
+HashAlgorithm::setStatus(const string& name, unsigned int port, bool status, time_t time, uint64_t latencyMs)
+{
+  TSDebug(DEBUG_TAG_INIT, "HashAlgorithm::setStatus name=%s status=%d", name.c_str(), status);
+  
+  HashNode* node = findStatusByNameAndPort(name,port);
+  if(node) {
+    node->setStatus(status,time,latencyMs);
+  } else {
+    TSError("Carp internal error setStatus host %s not found",name.c_str());
+  }
+}
+
+void
+HashAlgorithm::setStatus(HashNode * node, bool status, time_t time, uint64_t latencyMs)
+{
+  TSDebug(DEBUG_TAG_INIT, "HashAlgorithm::setStatus name=%s status=%d", node->name.c_str(), status);
+
+//  HashNode* node = findStatusByNameAndPort(name,port);
+  if(node) {
```

Comment:
Why is `node` checked if it's passed in? That seems more like an assert or shouldn't be checked at all.